source/auth/tests/legacy/connection-string.yml

```diff
@@ -412,6 +412,28 @@ tests:
     mechanism_properties:
       ENVIRONMENT: azure
       TOKEN_RESOURCE: foo
+- description: should accept a url-encoded TOKEN_RESOURCE (MONGODB-OIDC)
+  uri: mongodb://user@localhost/?authMechanism=MONGODB-OIDC&authMechanismProperties=ENVIRONMENT:azure,TOKEN_RESOURCE:mongodb%3A//test-cluster
```

Comment:
Connection strings can have options, and options have values. The option `authMechanismProperties` has a "[value that represents one or more key and value pairs](https://github.com/mongodb/specifications/blob/44a7ecfc6782b247ad75dcfd17c6bb4f76214f24/source/connection-string/connection-string-spec.md?plain=1#L218C20-L218C76)". That spec says that if a key contains a `,` or a `:`, it must be url-encoded (so should `%` and `&`,  though these are omitted). Here is an example where I url-encode the `TOKEN_RESOURCE` value `abc,def:ghi`, as per the connection string spec, which I understand this PR mirrors:
```
ConnectionString cs = new ConnectionString(
        "mongodb://localhost/?authMechanism=MONGODB-OIDC&authMechanismProperties="
                + "ENVIRONMENT:azure,TOKEN_RESOURCE:abc%2Cdef%3Aghi");
assertEquals("abc,def:ghi", cs.getCredential().getMechanismProperty("TOKEN_RESOURCE", null));
Expected :abc,def:ghi
Actual   :abc
```
The result is incorrect. When I place the same uri into a legacy `connection-string.json` test, it also fails, producing `abc`. As per "the values in connection options MUST be URL decoded by the parser" at the top of that section in the spec, Java decodes the entire value of the `authMechanismProperties` option, before parsing the name-value properties (the values are not url-decoded a second time). The spec lists Java as a reference implementation, along with Python, which I expect does the same.

This is an abstract problem in our connection string spec, but a concrete problem here because the token resource value appears to allow arbitrary characters. Documenting that the user must url-encode both the entire `authMechanismProperties` value, and the token resource (which ends up double-encoded) seems untenable. The only other (marginally less-bad) option I can think of is to specify that the `TOKEN_RESOURCE` value must always be base64 encoded - we would base64-decode, and then url-encode when sending to GCP.

```suggestion
  uri: mongodb://user@localhost/?authMechanism=MONGODB-OIDC&authMechanismProperties=ENVIRONMENT:azure,TOKEN_RESOURCE:abc%2Cdef%3Aghi
```

Reply:
I think it is a mistake to decode the full `authMechanismProperties`.  It has a defined structure, split on `:` and `,`.  Then within that you have your keys and values.

Reply:
We have the same behavior as @katcharov described here on C# driver. Will check how complex would be to change it.

Reply:
I suspect the intent was to mirror how URL properties behave, and those can always be url-encoded (and it's a reasonable practice to do so), so users may have an expectation that our connection strings work the same way. This clearly gets in the way of using certain values. 

However, changing this behaviour would break backward compatibility. For example, users might be encoding `readPreferenceTags=dc%3Any%2Crack%3A1` as per our spec, with the expectation that this would decode to `dc:ny,rack:1`.

We discussed a bit in Java, and our proposed plan is to disable full-decoding of `authMechanismProperties` only when `authMechanism=MONGODB-OIDC`, and in just that case decode individual values. This is in line with your proposal, but avoids breaking backward compatibility. I believe this should be clarified in the spec.

The rejected alternatives were to A) forbid `,` in `TOKEN_RESOURCE` (no, because this can be a connection string), or B) base32 encode the value (contrary to the Python release), or C) break spec.